Plutarch/TermCont.hs

```diff
@@ -0,0 +1,49 @@
+{-# OPTIONS_GHC -Wno-unused-foralls #-}
+
+module Plutarch.TermCont (
+  hashOpenTerm,
+  TermCont (TermCont),
+  runTermCont,
+  unTermCont,
+  tcont,
+) where
+
+import Data.Kind (Type)
+import Data.String (fromString)
+import Plutarch.Internal (Dig, PType, S, Term (Term), asRawTerm, getTerm, hashRawTerm)
+import Plutarch.Trace (ptraceError)
+
+newtype TermCont :: forall (r :: PType). S -> Type -> Type where
+  TermCont :: forall r s a. {runTermCont :: ((a -> Term s r) -> Term s r)} -> TermCont @r s a
```

Comment:
TIL!

What is this syntax? I had no idea GHC supported implicit type parameter on (new)types, that you can pass explicitly using `TypeApplications`.

Reply:
It's just GADT syntax with `newtype`.

Reply:
I'm also learning so much just by observing Plutarch :D

Reply:
For clarity, and foresaking newtype, this is equivalent to:

```haskell
type TermCont :: forall (r :: PType). S -> Type -> Type
data TermCont s a where
  TermCont :: forall r s a. {runTermCont :: ((a -> Term s r) -> Term s r)} -> TermCont @r s a
```

It is the `TermCont @r` part that I had no idea could be done in GHC; ie. implicit type parameters. The GADT unified the two `r`'s, which is a neat trick. So we are using GADT to workaround lack of dependent types.

Reply:
Essentially yeah this is the idea.